src/merkle/partial_mt/mod.rs

```diff
@@ -74,6 +78,80 @@ impl PartialMerkleTree {
         })
     }
 
+    /// Returns a new [PartialMerkleTree] instantiated with leaves map as specified by the provided
+    /// entries.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - If the depth is 0 or is greater than 64.
+    /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
+    /// - The provided entries contain an insufficient set of nodes.
+    pub fn with_leaves<R, I>(entries: R) -> Result<Self, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (NodeIndex, RpoDigest)> + ExactSizeIterator,
+    {
+        // a map where the key is the depth of the node, and the value is the vector of indexes and
+        // hashes of nodes with this depth
+        let mut layer_nodes: BTreeMap<u8, Vec<(u64, RpoDigest)>> = BTreeMap::new();
+
+        let mut leaves = BTreeSet::new();
+        let mut nodes = BTreeMap::new();
+
+        for (node_index, hash) in entries.into_iter() {
+            leaves.insert(node_index);
+            nodes.insert(node_index, hash);
+            layer_nodes
+                .entry(node_index.depth())
+                .and_modify(|layer_vec| layer_vec.push((node_index.value(), hash)))
+                .or_insert(vec![(node_index.value(), hash)]);
+        }
```

Comment:
Now that I think of it, `layer_nodes` probably can be even simpler as we already put the hash into the `nodes`. So, maybe having it be `BTreeMap<u8, Vec<u64>>` is enough?